iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,102 @@
+use std::borrow::BorrowMut;
+use std::cmp::Ordering;
+use std::future::Future;
+use std::io::{Error, ErrorKind};
+use std::net::SocketAddr;
+use std::ops::{Deref, DerefMut};
+use std::sync::Arc;
+
+use async_stream::stream;
+use futures::Stream;
+use tokio::io;
+use tokio::io::{AsyncReadExt, AsyncWriteExt, AsyncRead};
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::net::TcpStream;
+use tokio::sync::Mutex;
+use tokio::time::Instant;
+
+use iroha_actor::{Actor, Context, Handler, MessageResponse};
+use iroha_crypto::{Hash, PublicKey};
+use iroha_logger::{debug, error, info, trace, warn};
+
+use crate::message::{Message, MessageResult};
+use crate::peer_state::State;
+use iroha_data_model::prelude::PeerId;
+use crate::network::SendPeerMsg;
+
+pub struct Peer {
+    id: PeerId,
+    write: OwnedWriteHalf,
+    read: Option<OwnedReadHalf>,
+}
+
+impl Peer {
+    pub fn new(id: PeerId, stream: TcpStream) -> Self {
+        let (read, write) = stream.into_split();
+        Self { id, write, read: Some(read) }
+    }
+
+    fn stream(&mut self) -> impl Stream<Item = Result<MessageResult, std::io::Error>> {
```

Comment:
Can't make this method compatible with `ctx.notify_with(self.stream());` on line 59 :(